Initial documentation on Multiple Control Planes (#7035)
istio/istio.io: content/en/blog/2020/multiple-control-planes/index.md

---

> title: "Multiple Control Planes"

Reviewer: @howardjohn - may I suggest the title "Immutable Control Planes". This is groundbreaking work and should not be confused with multicluster and the pile of terms we have in there.

Plus @oaktowner will have something good to work with during the release. This may require a little rework on this document.

Author: I don't have any strong opinions on what we call this. It might make sense as "Control Plane Revisions" since that is how it is configured. Or for a title something like "Immutable Control Planes: introducing revisions" or something like that. Naming things is not my expertise, so I'll defer to others :slightly_smiling_face:

Reviewer 2: Well, given that the blog is talking about canary deployment of a control plane, the title should probably be something like:
```
Safely Upgrade Istio using a Canary Control Plane Deployment
```

> title: "Multiple Control Planes"

Author: ```suggestion
title: "Safely Upgrade Istio using a Canary Control Plane Deployment"
```

subtitle: Simplifying Istio upgrades by offering safe canary deployments of the control plane
description: Simplifying Istio upgrades by offering safe canary deployments of the control plane.
publishdate: 2020-05-19
attribution: "John Howard (Google)"
keywords: [install,upgrade,revision,control plane]
---

Canary deployments are a core feature of Istio. Users rely on Istio's traffic management features to safely control the rollout of new versions of their applications, while making use of Istio's rich telemetry to compare the performance of canaries. However, when it came to upgrading Istio, there was not an easy way to canary the upgrade, and due to the in-place nature of the upgrade, issues or changes found affect the entire mesh at once.

> Istio 1.6 will support a new upgrade model to safely canary new versions of Istio. In this new model, proxies will associate with a specific control plane that they use. This allows a new version to deploy to the cluster with less risk -- no proxies connect to the new version until the user explicitly chooses to. This allows gradually migrating workloads to the new control plane, while monitoring changes using Istio telemetry to investigate any issues - just like using `VirtualService` for workloads. Each independent control plane is referred to as a "revision" and has an `istio.io/rev` label.

Reviewer 2: ```suggestion
Istio 1.6 will support a new upgrade model to safely canary-deploy new versions of Istio. In this new model, proxies will associate with a specific control plane that they use. This allows a new version to deploy to the cluster with less risk -- no proxies connect to the new version until the user explicitly chooses to. This allows gradually migrating workloads to the new control plane, while monitoring changes using Istio telemetry to investigate any issues - just like using `VirtualService` for workloads. Each independent control plane is referred to as a "revision" and has an `istio.io/rev` label.
```

> In Istio 1.6, we will support a new upgrade model to safely canary new versions of Istio itself. In this new model, proxies will associate with a specific control plane that they want to use. This allows a new version to deploy to the cluster with much lower risk -- no proxies will connect to the new version until the user explicitly chooses to. This allows gradually migrating workloads to the new control plane, while monitoring changes using Istio telemetry to investigate any issues - just like using `VirtualService` for workloads. Each independent control plane is referred to as a "revision" and has an `istio.io/rev` label.

Reviewer 3: ```suggestion
Istio 1.6 will support a new upgrade model to safely canary new versions of Istio. In this new model, proxies will associate with a specific control plane that they use. This allows a new version to deploy to the cluster with less risk -- no proxies connect to the new version until the user explicitly chooses to. This allows gradually migrating workloads to the new control plane, while monitoring changes using Istio telemetry to investigate any issues - just like using `VirtualService` for workloads. Each independent control plane is referred to as a "revision" and has an `istio.io/rev` label.
```

## Understanding upgrades

Upgrading Istio is a complicated process. During the transition period between two versions, which might take a long time for large clusters, there are version differences between proxies and the control plane. In the old model the old and new control planes use the same Service, traffic is randomly distributed between the two, offering no control to the user. However, in the new model, there is not cross-version communication. Look at how the upgrade changes:

<iframe src="https://docs.google.com/presentation/d/e/2PACX-1vSbj4B52oEtQ8wGvmaSy29Zao3Q8Ex-w6JaripuJThMTK4F4bxDZkyNUSaexz8Rp8v4QCuDB2dAZkrv/embed?start=false&loop=true&delayms=3000" frameborder="0" width="960" height="569" allowfullscreen="true" mozallowfullscreen="true" webkitallowfullscreen="true"></iframe>

## Configuring

> Control plane selection is done based on the sidecar injection webhook. Each control plane is configured to select objects with a matching `istio.io/rev` label on the namespace. These pods will then be configured to connect to a control plane specific to that revision. Unlike in the current model, this means that a given proxy connects to the same revision during its lifetime. This avoids subtle issues that might arise when a proxy switches what control plane it is connected to.

Author: Basically how it works is when a pod is created, it will match a sidecar injector. Each revision has its own webhook, watching namespaces with the respective istio.io/rev label. So one of those will match the namespace, and the pod will be injected by that control plane. It will inject configuration to tell the proxy to connect back to that same control plane.

I want to avoid those implementation details, but I thought explaining here might help us come to a better explanation

> Control plane selection is done based on the sidecar injection webhook. Each control plane is configured to select objects with a matching `istio.io/rev` label on the namespace. These pods will then be configured to connect to a control plane specific to that revision. Unlike in the current model, this means that a given proxy connects to the same revision during its lifetime. This avoids subtle issues that might arise when a proxy switches what control plane it is connected to.

Author: And in the old model a proxy would switch which control plane it connects to during upgrades, since its upgrading in place. So the proxy has no idea, it just connects to "istiod", but the deployment behind "istiod" changes
